src/jdk.incubator.jextract/share/classes/jdk/incubator/jextract/tool/ConstantHelper.java

```diff
@@ -0,0 +1,476 @@
+                        MemoryAddress.class,
+                        LibraryLookup[].class,
+                        String.class)
+        );
+        this.MH_makeCString = findRuntimeHelperBootstrap(
+                runtimeHelper,
+                "makeCString",
+                methodType(
+                        MemoryAddress.class,
+                        String.class)
+        );
+
+        this.LIBRARIES = librariesDesc(findRuntimeHelperBootstrap(
+                runtimeHelper,
+                "libraries",
+                methodType(
+                        LibraryLookup[].class,
+                        String[].class)
+        ), libraryNames);
+
+        cw.visit(V15, ACC_PUBLIC, internalClassName, null, INTR_OBJECT, null);
+    }
+
+    private static DirectMethodHandleDesc findRuntimeHelperBootstrap(ClassDesc runtimeHelper, String name, MethodType type) {
+        return MethodHandleDesc.ofMethod(
+                Kind.STATIC,
+                runtimeHelper,
+                name,
+                desc(type)
+        );
+    }
+
+    public DirectMethodHandleDesc addLayout(String javaName, MemoryLayout layout) {
+        return emitCondyGetter(javaName + "$LAYOUT", MemoryLayout.class, desc(layout));
+    }
+
+    public DirectMethodHandleDesc addVarHandle(String javaName, String nativeName, MemoryLayout layout, Class<?> type, MemoryLayout parentLayout) {
+        return emitCondyGetter(javaName + "$VH", VarHandle.class, varHandleDesc(javaName, nativeName, layout, type, parentLayout));
+    }
+
+    public DirectMethodHandleDesc addMethodHandle(String javaName, String nativeName, MethodType mtype, FunctionDescriptor desc, boolean varargs) {
+        return emitCondyGetter(javaName + "$MH", MethodHandle.class, methodHandleDesc(nativeName, mtype, desc, varargs));
+    }
+
+    public DirectMethodHandleDesc addAddress(String javaName, String nativeName) {
+        return emitCondyGetter(javaName + "$ADDR", MemoryAddress.class, globalVarAddressDesc(nativeName));
+    }
+
+    public DirectMethodHandleDesc addFunctionDesc(String javaName, FunctionDescriptor fDesc) {
+        return emitCondyGetter(javaName + "$DESC", FunctionDescriptor.class, desc(fDesc));
+    }
+
+    public DirectMethodHandleDesc addConstant(String name, Class<?> type, Object value) {
+        if (type == MemoryAddress.class) {
+            if (value instanceof Long) {
+                return emitCondyGetter(name, type, addressDesc((Long) value));
+            } else if (value instanceof String) {
+                return emitCondyGetter(name, type, cStringDesc((String) value));
+            } else {
+                throw new IllegalStateException("Unhandled constant value type: " + value.getClass());
+            }
+        } else if (type.isPrimitive()) {
+            if (type == int.class || type == byte.class || type == short.class || type == char.class) {
+                return emitGetter(name, type, mv -> emitConInt(mv, ((Long) value).intValue()));
```

Comment:
I think having `emitIntConstantGetter`, `emitFloatConstantGetter` could be clearer. I noted that emitGetter is always used either through the condy case (which is fine) or through one of those constants getters here - I think *hiding* the lambda might make the code more readable here (and then you can keep existing general `emiteGetter` as an helper that is not to be used directly).